Mlem/Models/Content/Post/Post2Providing.swift

```diff
@@ -40,4 +43,117 @@ extension Post2Providing {
     var myVote_: ScoringOperation? { post2.myVote }
     
     var score: Int { upvoteCount - downvoteCount }
+    
+    func vote(_ newVote: ScoringOperation) {
+        if let oldTask = post2.tasks.vote, !oldTask.isCancelled { oldTask.cancel() }
+        post2.tasks.vote = Task(priority: .userInitiated) { await voteTask(newVote) }
+    }
+    
+    func voteTask(_ newVote: ScoringOperation) async {
+        if newVote == myVote { return }
+        
+        let oldVote = myVote
+        let oldUpvoteCount = upvoteCount
+        let oldDownvoteCount = downvoteCount
+        let oldReadStatus = isRead
+        
+        DispatchQueue.main.async {
+            self.upvoteCount += newVote.upvoteValue - oldVote.upvoteValue
+            self.downvoteCount += newVote.downvoteValue - oldVote.downvoteValue
+            self.myVote = newVote
+            self.isRead = true
+        }
+        
+        do {
+            let response = try await source.api.voteOnPost(id: id, score: newVote)
+            if !Task.isCancelled {
+                DispatchQueue.main.async {
+                    self.post2.update(with: response.postView, excludeActions: true)
+                    self.post2.tasks.vote = nil
+                }
+            } else {
+                print("\(newVote) task cancelled (Request)")
+            }
+        } catch ApiClientError.cancelled {
+            print("\(newVote) task cancelled (APIClient)")
+            post2.tasks.vote = nil
+        } catch {
+            print("\(newVote) task error: \(error)")
+            DispatchQueue.main.async {
+                self.myVote = oldVote
+                self.upvoteCount = oldUpvoteCount
+                self.downvoteCount = oldDownvoteCount
+                self.isRead = oldReadStatus
+            }
+            post2.tasks.vote = nil
+        }
+    }
+    
+    func toggleSave() {
+        if let oldTask = post2.tasks.save, !oldTask.isCancelled { oldTask.cancel() }
+        post2.tasks.save = Task(priority: .userInitiated) { await toggleSaveTask() }
+        if myVote != .upvote, UserDefaults.standard.bool(forKey: "upvoteOnSave") {
+            vote(.upvote)
+        }
+    }
```

Comment:
Upvote still applies on unsave--this should fix it 

```suggestion
    func toggleSave() {
        let wasSaved = isSaved
        if let oldTask = post2.tasks.save, !oldTask.isCancelled { oldTask.cancel() }
        post2.tasks.save = Task(priority: .userInitiated) { await toggleSaveTask() }
        if myVote != .upvote, !wasSaved, UserDefaults.standard.bool(forKey: "upvoteOnSave") {
            vote(.upvote)
        }
    }
```

Reply:
Oops, I forgot to commit that. Thanks :D